### DISCUSSION TITLE
Help troubleshooting F7 Lightning Plus

### DISCUSSION THREAD

**Discussion by Angelworks** [keirf/greaseweazle#74] [Q&A] [answered]
I assembled one of these from scratch - I included a gallery of pictures. I think I did ok at putting it together, but during the flash step it says I can't connect, and the st-link programmer just flashes yellow (connection error).

I included some detailed pictures from my scope - I'm guessing maybe I got a part backwards or something (although I've double checked a bunch of times) I'm using an official ST-link V2 Programmer, but I do have a cheapy knockoff coming from China in case this one is having issues.

Sorry for the picture spam ;). I guess I just need some pointers on where to start troubleshooting. I admit I'm not super well versed on microcontrollers - I've had a rough time just finding a pinout for this one. I did just notice U5 is missing some solder, easy to fix, but that shouldn't affect programming right?

![PXL_20210130_031109215](https://user-images.githubusercontent.com/19831611/106345939-565d2d80-6268-11eb-854f-a9b96686ef05.jpg)
![2019_0706_051024_001](https://user-images.githubusercontent.com/19831611/106345960-65dc7680-6268-11eb-92ac-bc95498a4d6e.JPG)
![2019_0706_051044_002](https://user-images.githubusercontent.com/19831611/106345961-67a63a00-6268-11eb-9846-68e346b6fa21.JPG)
![2019_0706_051105_003](https://user-images.githubusercontent.com/19831611/106345962-683ed080-6268-11eb-9291-83fb29a86434.JPG)
![2019_0706_051118_004](https://user-images.githubusercontent.com/19831611/106345964-68d76700-6268-11eb-8189-4df7bc76f063.JPG)
![2019_0706_051140_005](https://user-images.githubusercontent.com/19831611/106345965-6a089400-6268-11eb-969f-6886bd70ca37.JPG)
![2019_0706_051156_006](https://user-images.githubusercontent.com/19831611/106345967-6b39c100-6268-11eb-8e07-8075bbcafba5.JPG)

**Comment by wohali** [comment]
It may help to power the board separately and not rely on the STLink to power the board while you program it.

Looks like maybe shorts on R12 and R7? Can't be sure.

It's tedious, but check for any continuity between adjacent pins on the STM. Yes, all 100.

**Reply by keirf** [reply]
Good general advice. The potential shorts on the lower rows of R7 and R12 don't matter here because all those pins across all three arrays are connected to 3v3. Hence they're electrically connected anyway.

**Reply by Angelworks** [reply]
I noticed that as well, but there's actually a trace between those two pins so I think its ok (that side is basically one big plane anyhow).

I'll test every adjacent pin on the microcontroller as well. Visually (under the microscope) they all look perfect. I did test continuity between the chip and the pad for each pin as well - every appears to be soldered on there.

**Comment by keirf** [comment, marked as answer]
The big gotcha here is that the official ST-Link does not provide power. Its VCC pin is an input to power its level shifters, so it can work safely with different voltage levels (eg, 1.8v, 3.3v). So, you need to provide external power or, more easily, just hook up your ST-Link clone because those **DO** provide 3v3 and 5v power. Just make sure to hook up the 3v3 ;)

I should have mentioned this big gotcha in the wiki. That's now fixed.

Your soldering looks great, but a couple of points:
1. U5 pin 14 isn't soldered!
2. XO1 is the wrong way round

Be very careful messing with XO1. If you get stray solder on the package I've found they no longer work. Hot air it off, and I would probably hot air it back on again with more flux.

**Reply by Angelworks** [reply]
Yeah I noticed U5 wasn't soldered after taking the picture :) - was kinda surprised I missed that :).

X01 - I flipped it around using hot hair, but still had problems - I noticed that the pad in the upper right has continuity to the metal part on the top - so I've likely damaged it. It did have a bit of solder in that gap between the top and the center. Oh well lesson learned. I've ordered some more from Digikey.

Thank you for the advice!

**Reply by keirf** [reply]
With XO1 missing you can still get on with programming the device. In fact I program them immediately after soldering the MCU, with no other components at all.

You should then find that the status/act led flashes three times, in a loop. This is an error indicator that the external oscillator didn't start up.

**Reply by Angelworks** [reply]
Hmm I still can't get it to work. I did check that the swclk and swdio have continuity to the header pins. I've checked adjacent pins as well for shorts, but haven't been able to find any. Maybe it's my programmer?

**Reply by wohali** [reply]
For that programmer, I find it helps to set the "Connect Under Reset" option in the official STM32 ST-Link program. This selects the Hardware Reset reset mode. You can also press the "Refresh" button on that dialog box to be sure the programmer detects the chip - it should see the target P/N and the voltage of the device as well.

Here's the pinout mapping I use to program with that device. Note the use of Pin 1, which is the sense pin for the ST-Link to check the voltage of the device under test - it uses an ADC for this.


```
=============================
| 2                      20 |
| 1     7 9              19 |
=============================

DO NOT CONNECT USB TO GW DURING PROGRAMMING!!!

1: Target VCC (on F7 v1, BOOT, on the right-most connector nearest USB connector)
7: DIO
9: CLK
19: VDD (3.3V)
20: GND

STM32 ST-LINK Settings (Windows only)
* SWD 4MHz
Connect Under Reset
Reset Mode Hardware Reset
```

With this setup, it will only see ~2.9V for the board being programmed, because of the pullup for the BOOT pin. This doesn't seem to affect programming, but if you're worried, you can use a breadboard to connect both pins 1 and 19 on the ST-Link to VDD (3.3V).

**Reply by keirf** [reply]
Ah does pin 19 supply 3v3? I hadn't realised. The diagrams I've seen show that pin (and a lot of others) N/C. Is there a more comprehensive documentation of the ST-Link v2 pinout somewhere?

Connect Under Reset is a good idea.

This is why I hold header pins in place on the PCB immediately after soldering the MCU and program it straight away. That's enough variables for me, without soldering the bird seed too. I should probably suggest this as a tip on the wiki page.

**Reply by wohali** [reply]
Yup, you can power the circuit with it. Convenient, isn't it? It doesn't have a lot of current, but it seems to be enough for the GW F7 at least. (Strictly speaking, the manual says you shouldn't connect it. I haven't had a chance to test with a Lightning Plus yet.)

Search for UM1075 on this page under documentation for the best pinout description: https://www.st.com/en/development-tools/st-link-v2.html It's on page 12.

As for soldering, I reflow the boards in an oven here, so it's all or nothing. No problems so far on 20+ units.

**Reply by Angelworks** [reply]
Woot! It connects, and I see the memory and the device id. When I go to write and verify I get an internal command error, but it actually detects the mcu now!

I tested the vdd pin (1) and its at 1.7 volts.

![image](https://user-images.githubusercontent.com/19831611/106398888-478e8c00-63ca-11eb-8ca7-770b5d027e24.png)
![image](https://user-images.githubusercontent.com/19831611/106398913-73aa0d00-63ca-11eb-80e5-bfb112418e8b.png)

**Reply by keirf** [reply]
I would try and separately power the Greaseweazle rather than use the ST-Link power. For example, just connect a USB lead to the USB-B connector, with the PWR_SEL jumpered to USB. And disconnect 3.3v from the programming header!

**Reply by Angelworks** [reply]
You know what ultimately "solved it" was a work-around. I found that the programmer didn't like it when the device was powered by usb or bench power supply (for whatever reason). Running a jumper between 3.3v and acc (pin 1) worked around this - and it works!

Thank you everyone for your patience and help :).

**Reply by keirf** [reply]
Oh yes pin 1 on the programmer has to be fed target voltage:
1. The software checks for a valid voltage
2. It powers level shifters inside the programmer

So yes jumpering 1 to 19 is fine! Or you could have powered from USB, and connected pin 1 (not 19!) to 3.3v on the board.

Anyway, problem solved. Hopefully your board flashes the ACT LED three times now? Just waiting for the external XO.

**Reply by Angelworks** [reply]
Yup - it just flashes 3x, but I'll be able to fix the crystal tomorrow (I think). I'll just use hot air to attach it to the pcb this go so I don't end up grounding the top part to the bottom part.

**Comment by Angelworks** [comment]
I thought I'd follow up :) - it works perfectly fine now. I installed a new crystal using just hot air today. I just need to learn the command line options and all the associated utilities:

https://twitter.com/Angelworks42/status/1357133005332963328 (some old Samsung 3.5" floppy).

Edit: Oh and too add thank you very much again for everyone's patience and help :)!
